### DISCUSSION TITLE
Provide a way to opt-out middlewares from running at SSG

### DISCUSSION THREAD

**Discussion by danielo515** [withastro/roadmap#867] [Proposal]
### Body

Here we go again...

# Summary

I want to be able to opt out of middlewares running at build step (SSG phase)

# Background & Motivation

I use middlewares to do both authentication and authorization, which does not make sense for static pages, so in the best case they are just overhead, and in the worst they prevent some pages from being properly rendered.

# Goals

- I would love to have a mechanism to opt out from middleware executiong during build phase
- Ideally being able to decide which ones to run by exporting some env variable like `process.env.SSG` or something like that so I can decide in the middleware based on that.

# Example

Something like I already mentioned `import.meta.env.SSG`

**Comment by ematipico** [comment]
@danielo515 please fill the template

**Reply by danielo515** [reply]
WTF? I filled the template, don't understand how did it ended like that

**Comment by Fryuni** [comment]
Just found this. It is already possible to differentiate between those stages, but it is not intuitive if you are unfamiliar with Astro details (maybe even some internals).

So, I just published that as an integration to make it easy to use: https://inox-tools.vercel.app/astro-when

Specifically, you might want to look at the middleware example: https://github.com/Fryuni/inox-tools/blob/main/examples/astro-when/src/middleware.ts

---

This is not to say that this won't be available as a higher-level API part of Astro, nor that it will. I'm just providing a wrapper on how it is already currently possible with the lower-level APIs. The integration above should continue to work even if Astro gets its own API for it.

**Comment by IndependentCreator** [comment]
Related - I've configured my (Astro 4.10.1) project with output mode _server_ and my middleware.ts file code is checking context.request.headers. When I configure a specifc astro page with `export const prerender = true;` I see this warning in the terminal at runtime:

```
[WARN] `Astro.request.headers` is unavailable in "static" output mode, and in prerendered pages within "hybrid" and "server" output modes. If you need access to request headers, make sure that `output` is configured as either `"server"` or `output: "hybrid"` in your config file, and that the page accessing the headers is rendered on-demand.
```

**Reply by ematipico** [reply]
That's expected. Prerendered pages will be built, and they will become HTML pages, so it's expected that you won't have access to `Astro.request.headers`.

**Reply by IndependentCreator** [reply]
> That's expected. Prerendered pages will be built, and they will become HTML pages, so it's expected that you won't have access to `Astro.request.headers`.

That was my understanding, but this means that if we add code to middleware that accesses `request.headers` we will get this warning in the terminal for every request for a `prerender = true` page. Is there any way to supress the warnings?

**Reply by ematipico** [reply]
We put in place that warning for this particular reason.

I'm not really sure what your use case is, but you should consider other tools instead of request headers, for example `Astro.locals`

**Reply by IndependentCreator** [reply]
Thanks for your help. In my case, this line from middleware.ts triggers the warnings for `prerender = true` pages:
`const sessionId = context.cookies.get( "mycookiename" );`

I'm trying to follow a pattern recommended by _lucia-auth_: checking for a session cookie in middleware.ts, retrieving the corresponding session and user from the database, and saving the session and user variables in `Astro.locals` so that I can access them from the `prerender = false` pages.

I was thinking that some pages won't need session information and therefore they could be set to `prerender = true`, but that triggers the warnings. I'm open to suggestions if there's a pattern better suited to doing this in Astro.

**Reply by ematipico** [reply]
If you think about it, if you don't need session information, hence `const sessionId = context.cookies.get( "mycookiename" )` isn't needed.

**Reply by IndependentCreator** [reply]
Yes, but if I add `const sessionId = context.cookies.get( "mycookiename" )` to middleware.ts, it gets run for every page, even the (`prerender = true`) ones where it doesn't function and it isn't needed, correct?

So the only ways to avoid the warning would be A) use `prerender = false` for all pages, regardless of whether they need session information, or B) Only call `const sessionId = context.cookies.get( "mycookiename" )` from `prerender = false` pages (instead of calling it from middleware.ts). Am I understanding correctly?

**Reply by ematipico** [reply]
Oooh, I see! That makes sense now! I'll think about it

**Reply by IndependentCreator** [reply]
In the future it might be nice to have 2 separate hooks, e.g. `defineMiddlewarePrerenderTrue()` and `defineMiddlewarePrerenderFalse()`

In my case I have refactored my code so that all of the session cookie retrieval happens inside my SessionData class. For my specific situation this is probably more natural, because the SessionData class was already needed for precisely those pages that require session access 🙂

**Reply by ryuheechul** [reply]
That's a good idea and after tinkering a bit on that idea, I was able to create the `defineMiddleWarePrerenderFalse` like this which essentially is a wrapper and seems to be working for me.
```ts
const defineMiddlewarePrerenderFalse: typeof defineMiddleware = (fn) => {
  return defineMiddleware((context, next) => {
    if (context.isPrerendered) {
      return next();
    }

    return fn(context, next);
  });
};

// which works in the same style as the `defineMiddleware`
export const onRequest = defineMiddlewarePrerenderFalse(async (context, next) => {
  // but anything here will be ignored if prerendered
});
```

**Comment by danielo515** [comment]
Yes, but the middleware running is the unexpected thing

**Reply by ematipico** [reply]
I am sure you come from a different background, so I understand what you mean. However, we document its behaviour in the documentation, so there's no unexpected things: https://docs.astro.build/en/guides/middleware/

> Middleware allows you to intercept requests and responses and inject behaviors dynamically every time a page or endpoint is about to be rendered. This rendering occurs at build time for all prerendered pages, but occurs when the route is requested for pages rendered on demand.

**Comment by danielo515** [comment]
Just to be sure I understand, can someone explain me what is the use case
for a middleware running in static generation? They are inherently dynamic
to me

**Reply by Fryuni** [reply]
Things like redacting, authenticating server-side tools that are per page, any shared logic between routes, populating the `Astro.locals` with information that needs to be accessible by components deep in the page without prop drilling, etc.

**Comment by delucis** [comment]
As [mentioned above](https://github.com/withastro/roadmap/discussions/867#discussioncomment-12783336) it is already possible for middleware to execute its logic conditionally so it doesn't apply to prerendered pages.

```js
    if (context.isPrerendered) {
      return next();
    }
```

Given that, I don't we're interested in adding an additional API for this.

The current API is flexible and means you can even choose to only run parts of middleware for prerendered pages.

**Reply by danielo515** [reply]
I'm pretty sure that was not available, or at least not documented when I opened the issue one year ago.
Is it documented now?

**Reply by delucis** [reply]
Very possibly! Sorry, should maybe have phrased that as “is _now_ possible” — wasn’t paying attention to the original post date.

`isPrerendered` was added in Astro v5 and is documented here: https://docs.astro.build/en/reference/api-reference/#isprerendered

**Reply by danielo515** [reply]
Ah, wonderful. Many thanks for the reference. Indeed this was 4.x era, so this counts as "done" or "implemented".
Thanks again.
